### ISSUE TITLE
Run the Promises/A+ test suite in real browsers

### ISSUE THREAD

**Issue by mzgol** [jquery/jquery#2152]
We're currently running the Promises/A+ tests only in Node and only on `master`. That mean `compat` is mostly untested and on `master` we don't know if it really works in all the browsers.

We should get the test suite run in real browsers. IE8 will be a problem since the suite uses getters heavily but even getting it all run in other browsers would be a huge win.

It's not a must-have for the beta IMO but we should have it ready before 3.0.0 final so I'll set the milestone.

Refs gh-1722
Refs gh-1996

**Comment by mzgol** [comment]
Moving to 3.0.1.

**Comment by yanickrochon** [comment]
Does this mean that jQuery will actually, finally, have a Promise/A+ implementation??

**Comment by mzgol** [comment]
@yanickrochon Yup, it's already available to test in [jQuery 3.0.0-alpha1](https://code.jquery.com/jquery-3.0.0-alpha1.js) & [jQuery Compat 3.0.0-alpha1](https://code.jquery.com/jquery-compat-3.0.0-alpha1.js) (if you need IE8 support).

This issue is only about testing it in real browsers, so far we're testing it in Node with jsdom.

**Comment by guest271314** [comment]
@mgol Will to help. What needs to be done? How to achieve this? See also http://stackoverflow.com/questions/36438228/how-to-test-if-jquery-3-0-beta-is-promises-a-compatible-in-browser

**Comment by mgol** [comment]
A separate QUnit module needs to be defined where we have to wrap the Promises/A+ test suite so that we can have it run automatically in all browsers at http://swarm.jquery.org/project/jquery; en example run: http://swarm.jquery.org/job/2582. This might not be straightforward to achieve, I expect to have to apply some hackery to get it done properly so I wanted to get to it after we release 3.0.0.

Why do you ask? Would you like to try to achieve it or do you ask for your own test purposes?

**Comment by guest271314** [comment]
@mgol Would like to help achieve it. Have little to no experience with QUnit, though was able to run version 3.0.0-beta1 through QUnit yesterday, where all 28 tests of newly released deferred.js tests passed. How can this user help?

**Comment by guest271314** [comment]
@mgol Actually, it was jquery-git.js that ran at QUnit. Downloaded jQuery from GitHub. Had to include   	`<script src="data/testrunner.js"></script>` within `/jquery-master/test/index.html` and change source at `/jquery-master/test/jquery.js/` at `document.write( "<script id='jquery-js' src='/path/to/jquery-git.js'><\x2Fscript>" );`

**Comment by mgol** [comment]
Deferred tests don't contain the Promises/A+ one. A new test would have to be added to [Gruntfile.js](https://github.com/jquery/jquery/blob/5cbb234dd3273d8e0bbd454fb431ad639c7242c1/Gruntfile.js#L136-L164), to [test/data/testinit.js](https://github.com/jquery/jquery/blob/ce6c83f710c28108ccb4d50a7b924baa890dc961/test/data/testinit.js#L304-L332) and then a new module created inside [test/unit/](https://github.com/jquery/jquery/tree/master/test/unit). This file would need to define a QUnit test that would invoke the Promises/A+ test suite run in a browser. This will require writing a browser adapter for the test suite similar to [test/promises_aplus_adapter.js](https://github.com/jquery/jquery/blob/ce6c83f710c28108ccb4d50a7b924baa890dc961/test/promises_aplus_adapter.js), more information is available in the README of the [`promises-aplus-tests` npm package](https://www.npmjs.com/package/promises-aplus-tests). That should be mostly it.

**Comment by guest271314** [comment]
@mgol Have little to no experience using nodejs, and have not previously composed an adapter. Reviewed the README at https://www.npmjs.com/package/promises-aplus-tests and found https://tonicdev.com/npm/promises-aplus-tests. How can jQuery-3.0.0-beta1 Deferred be exposed and run there?

**Comment by guest271314** [comment]
@mgol  Tried below at https://tonicdev.com/npm/promises-aplus-tests

```
require("promises-aplus-tests");
require( "jsdom" ).env( "", function( errors, window ) {
	if ( errors ) {
		console.error( errors );
		return;
	}

	var jQuery = require( "https://code.jquery.com/jquery-3.0.0-beta1.js" )( window );

	exports.deferred = function() {
		var deferred = jQuery.Deferred();

		return {
			promise: deferred.promise(),
			resolve: deferred.resolve.bind( deferred ),
			reject: deferred.reject.bind( deferred )
		};
	};
} );
```

received error 

`SyntaxError: /app/available_modules/1460391167000/jsdom/lib/jsdom.js:3 () => `jsdom 7.x onward only works on Node.js 4 or newer: https://github.com/t ^ Unexpected token )`

**Comment by mgol** [comment]
You haven't replied - do you want to fix this issue in jQuery or just run tests for yourself? Tonicdev runs code in Node, not in a browser so that doesn't seem like it solves your willingness to have it run in a browser. And the error you got is caused because Tonicdev by default runs it in an old version of Node, 0.12.

**Comment by guest271314** [comment]
@mgol  

> You haven't replied - do you want to fix this issue in jQuery or just run tests for yourself? 

Yes, fix issue in jQuery. Attempted to convey this at 

> Would like to help achieve it.
> How can this user help?

**Comment by mgol** [comment]
OK. :) You'll need to set it up locally then, Tonicdev won't help you here. You need to set up the test environment (read https://github.com/jquery/jquery#running-the-unit-tests), run the QUnit test suite by building jQuery via Grunt (it's described in the README) and navigating to http://YOUR_LOCAL_SERVER_POINTING_TO_JQUERY_REPO/test.

If you don't have experience with Node & QUnit then I need to warn you, though - it doesn't seem extremely easy so I wouldn't classify it as a good first bug.

**Comment by guest271314** [comment]
@mgol  Only the https://github.com/jquery/jquery#running-the-unit-tests portion, or following instructions for entire page?

**Comment by mgol** [comment]
I think everything there may be useful for you, maybe except the `modules` section as you're be building the whole of jQuery or running it in AMD mode. In particular, you need to have a working Apache+PHP setup with recent Node.js.

**Comment by guest271314** [comment]
@mgol Can run http://php.net/manual/en/features.commandline.webserver.php locally. Re-installed nodejs yesterday. Not certain what a recent version is though.

**Comment by mgol** [comment]
A real web server is needed e.g. for AJAX tests. We also run JSHint, build the files from AMD to the one final file etc. so Node.js is needed. Now, we'd love to not have to use Apache+PHP but we have lots of tests using PHP and rewriting them would be a long, non-trivial task so we have what we have currently.

The built-in PHP server will fail AJAX tests. But this task doesn't require to run AJAX tests so it might be fine for you.

**Comment by guest271314** [comment]
@mgol Ok. Will review the procedures described at links. Not presently concerned with AJAX or other jQuery method tests; downloaded zip of jQuery yesterday and QUnit provided adequate results. When build jQuery at command-line will this include version 3.0.0-beta1, or will it be necessary to switch the build version for a copy of 3.0.0-beta1 after the build is complete?

**Comment by guest271314** [comment]
@mgol  Found requirebin, receiving error `/usr/lib/node_modules/browserify/node_modules/assert/assert.js:361Uncaught TypeError: fs.readdir is not a function` at http://requirebin.com/?gist=ce7a014d0ec638de5c13516d69a1e64e

**Comment by guest271314** [comment]
@mgol Tried at a different browser received this error `Uncaught TypeError: undefined is not a function /tmp/promises-aplus-tests11638-18974-ukg7u2/node_modules/promises-aplus-tests/lib/programmaticRunne…:79`

**Comment by guest271314** [comment]
@mgol Could not reproduce tests. See http://chat.stackoverflow.com/rooms/108948/discussion-between-jared-smith-and-guest271314. How did the jQuery team that ran the tests do so?

**Comment by guest271314** [comment]
@timmywil How did the team run the tests described at [jQuery 3.0 Beta Released ](http://blog.jquery.com/2016/01/14/jquery-3-0-beta-released/) ?

> jQuery.Deferred objects have been updated for compatibility with Promises/A+ and ES2015 Promises, verified with the [Promises/A+ Compliance Test Suite](https://github.com/promises-aplus/promises-tests).

**Comment by timmywil** [comment]
@guest271314 We included the test suite in the tests we run during the grunt task.

**Comment by guest271314** [comment]
@mgol https://github.com/guest271314/jquery-3.0.0-promises-tests-in-browser

**Comment by guest271314** [comment]
@mgol Doesapproach at link return expected results?

**Comment by mgol** [comment]
@guest271314 I'm not sure what you're asking for; could you elaborate?

**Comment by guest271314** [comment]
@mgol Does approach at http://guest271314.github.io/jquery-3.0.0-promises-tests-in-browser/ return expected results described at OP?

> We should get the test suite run in real browsers.

**Comment by mgol** [comment]
@guest271314 I described what needs to be done in https://github.com/jquery/jquery/issues/2152#issuecomment-208424557. Your proof of concept repository doesn't do all that, it doesn't even use QUnit.

**Comment by markelog** [comment]
Sorry for interrupting, we wanted to start using karma at some point, maybe this could be our first step? What do you think?

**Comment by mgol** [comment]
Perhaps. These tests *should be* stable enough to make it work at least in our main browsers. I'd expect problems with the Android ones, though and we don't have a nice story of recovering from such failed tests etc.

**Comment by markelog** [comment]
Yeah, probably, although there isn't a lot of tests for promise and we support only 4.0 as i recall? So we might just pass through

**Comment by guest271314** [comment]
@mgol Have no experience with `QUnit` other than running your `deferred.js`. The goal here was to run the Promises/A+ compliance tests in the browser, to test jQuery 3+ Deferred, however that could be achieved. What you mean by "doesn't do all that"? Are the tests not running correctly? Have no experience with `grunt` other than running `grunt test`. Willing to help, with clear guidance as to what to do. Are you saying that `QUnit` can run the tests? If yes, why hasn't the Promises/A+ compliance test been included into the files you referenced above, by the composers of the same files?
